### PR TITLE
Add install feature in CMake

### PR DESCRIPTION
**Pull request by bansan85**
Add feature to install header via `make install`.

I also change:

  - rename project from `rlbox-cpp17` to `rlbox-sandboxing-api`
  - install header are in `include/rlbox/rlbox*.hpp` and not simply `include/rlbox*.hpp`
  - add `find_package` feature

Example of find_package:

```
cmake_minimum_required(VERSION 3.10)
project(zpipe CXX)
find_package(rlbox-sandboxing-api CONFIG REQUIRED)
add_executable(zpipe main.cpp mylib.c)
target_link_libraries(zpipe rlbox-sandboxing-api)
```

This is for this feature that I propose to add an option to not build tests and avoid `catch2` dependency #28.

### CODE REVIEW

**Comment by shravanrn** [comment]
I'm reviewing this now

File: CMakeLists.txt [hunk 1]
```diff
@@ -158,17 +159,51 @@ endif()
...
+target_link_libraries(${PROJECT_NAME} INTERFACE ${CMAKE_THREAD_LIBS_INIT})
+
+# Install ###################
+
+include(CMakePackageConfigHelpers)
+write_basic_package_version_file("${PROJECT_NAME}ConfigVersion.cmake"
+                                 VERSION ${CMAKE_PROJECT_VERSION}
+                                 COMPATIBILITY SameMajorVersion)
+
+set_target_properties(${PROJECT_NAME} PROPERTIES PUBLIC_HEADER "${RLBOX_SOURCE_FILES}")
```
Comment (shravanrn) [inline CMakeLists.txt:179]: Maybe move this line into the "# Targets ###################" section

File: CMakeLists.txt [hunk 2]
```diff
@@ -158,17 +159,51 @@ endif()
...
+
+set(INCLUDE_INSTALL_DIR "${CMAKE_INSTALL_INCLUDEDIR}")
```
Comment (shravanrn) [inline CMakeLists.txt:181]: Can we get rid of this temporary variable and just inline the uses?

**Comment by shravanrn** [comment]
Thanks for submitting this! Looks good overall (couple of minor comments). I just want to test this in Windows & Linux & Mac before landing this. I'll try to do this over the next 24 hours.

File: CMakeLists.txt [hunk 3]
```diff
@@ -158,17 +159,51 @@ endif()
...
+set(DATAROOT_CONFIG_DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/cmake/${PROJECT_NAME}")
+configure_package_config_file(${PROJECT_SOURCE_DIR}/cmake/${PROJECT_NAME}Config.cmake.in
+  ${PROJECT_BINARY_DIR}/${PROJECT_NAME}Config.cmake
+  INSTALL_DESTINATION ${DATAROOT_CONFIG_DESTINATION}
+  PATH_VARS INCLUDE_INSTALL_DIR)
+install(TARGETS ${PROJECT_NAME}
```
Comment (shravanrn) [inline CMakeLists.txt:187]: The catch2 projects seem to run install only when they aren't included as a subdirectory due to bug https://github.com/catchorg/Catch2/issues/1373

See https://github.com/catchorg/Catch2/blob/devel/CMakeLists.txt#L109
We should probably do the same here

**Comment by shravanrn** [comment]
@bansan85 Could you post a test of how to use the cmake installed files. Here is what I have so far, but this doesn't seem to work

```
cmake_minimum_required(VERSION 3.13)

project(rlbox_cmake_test
        VERSION 0.1
        DESCRIPTION "RLBox rlbox_cmake_test")

find_package(rlbox-sandboxing-api)
if(NOT rlbox-sandboxing-api_FOUND)
    message(FATAL_ERROR "NO rlbox-sandboxing-api_FOUND")
endif()

add_executable(testt test.cpp)
target_link_libraries(testt rlbox-sandboxing-api)
```

File: CMakeLists.txt (hunk 1 above)
Comment (bansan85) [reply CMakeLists.txt]: Fix

File: CMakeLists.txt (hunk 2 above)
Comment (bansan85) [reply CMakeLists.txt]: Fix

File: CMakeLists.txt (hunk 3 above)
Comment (bansan85) [reply CMakeLists.txt:186]: I looked at the bug. I uninstall system catch2 to force local catch2.

After a:

```
make DESTDIR=/tmp/ttt install
cd /tmp/ttt/
find
```

I have:

```
.
./usr
./usr/local
./usr/local/share
./usr/local/share/cmake
./usr/local/share/cmake/rlbox-sandboxing-api
./usr/local/share/cmake/rlbox-sandboxing-api/rlbox-sandboxing-apiTargets.cmake
./usr/local/share/cmake/rlbox-sandboxing-api/rlbox-sandboxing-apiConfig.cmake
./usr/local/share/cmake/rlbox-sandboxing-api/rlbox-sandboxing-apiConfigVersion.cmake
./usr/local/rlbox
./usr/local/rlbox/rlbox_range.hpp
./usr/local/rlbox/rlbox_type_traits.hpp
./usr/local/rlbox/rlbox_sandbox.hpp
./usr/local/rlbox/rlbox_stdlib_polyfill.hpp
./usr/local/rlbox/rlbox_struct_support.hpp
./usr/local/rlbox/rlbox_stdlib.hpp
./usr/local/rlbox/rlbox_helpers.hpp
./usr/local/rlbox/rlbox_types.hpp
./usr/local/rlbox/rlbox_unwrap.hpp
./usr/local/rlbox/rlbox_noop_sandbox.hpp
./usr/local/rlbox/rlbox_wrapper_traits.hpp
./usr/local/rlbox/rlbox_policy_types.hpp
./usr/local/rlbox/rlbox.hpp
./usr/local/rlbox/rlbox_app_pointer.hpp
./usr/local/rlbox/rlbox_conversion.hpp
./usr/local/rlbox/rlbox_dylib_sandbox.hpp
```

So I do not have files from catch2.

According to the ticket from catch2, it appends only if `include(GNUInstallDirs)` is previous `add_subdirectory("${catch2_SOURCE_DIR}")`. It's not the case if rlbox.

I also tried to move `include(GNUInstallDirs)` before `find_package(Catch2 QUIET)` and I don't have files from catch2 neither.

Did you had this bug ?

**Comment by bansan85** [comment]
cmake example added :)

File: CMakeLists.txt (hunk 3 above)
Comment (shravanrn) [reply CMakeLists.txt:186]: Oh i meant a similar thing would happen if someone included the rlbox project as a sub_project and then ran make install in their project, then rlbox may get installed in a sub directory there (like catch does). Its not a big deal either way though --- i think the fix is to disable the install target if this is a sub_project as shown in the catch repo link

**Comment by shravanrn** [comment]
> cmake example added :)

Ah excellent :+1:! This looks great! I'll run a quick test on Windows/Mac/Linux in the next 24 hours, and we should be able to merge it soon

**Comment by bansan85** [comment]
Thanks.
About the workaround, I will try some tests to understand the bug. Maybe add a warning to detect if include(GNUInstallDirs) was already called before your call.

**Comment by bansan85** [comment]
I made a try for the bug about GNUInstallDirs with the following file:

```
cmake_minimum_required(VERSION 3.13)

project(rlbox_cmake_test
        VERSION 0.1
        DESCRIPTION "RLBox rlbox_cmake_test")

include(GNUInstallDirs)

add_subdirectory(rlbox_sandboxing_api)

add_library(lib mylib.c)

add_executable(testt main.cpp)
target_link_libraries(testt rlbox-sandboxing-api lib)


install(FILES ${CMAKE_CURRENT_BINARY_DIR}/testt
        DESTINATION
        ${CMAKE_INSTALL_DOCDIR}/${PROJECT_NAME}/TTT)
```

when I make a `make install`, I don't have problem with `${PROJECT_NAME}` but it also install header from rlbox:

```
-- Install configuration: "Release"
-- Installing: /tmp/ttt2/usr/local/share/doc/rlbox_cmake_test/rlbox_cmake_test/TTT/testt
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_app_pointer.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_conversion.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_dylib_sandbox.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_helpers.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_noop_sandbox.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_policy_types.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_range.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_sandbox.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_stdlib.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_stdlib_polyfill.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_struct_support.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_type_traits.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_types.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_unwrap.hpp
-- Up-to-date: /tmp/ttt2/usr/local/include/rlbox/rlbox_wrapper_traits.hpp
-- Up-to-date: /tmp/ttt2/usr/local/share/cmake/rlbox-sandboxing-api/rlbox-sandboxing-apiTargets.cmake
-- Up-to-date: /tmp/ttt2/usr/local/share/cmake/rlbox-sandboxing-api/rlbox-sandboxing-apiConfig.cmake
-- Up-to-date: /tmp/ttt2/usr/local/share/cmake/rlbox-sandboxing-api/rlbox-sandboxing-apiConfigVersion.cmake
```

Any way, I don't think it's a good idea to `add_subdirectory` a 3rd-party library.

**Comment by shravanrn** [comment]
Sorry took me a while to get to this. Confirmed that this works fine on Linux and Mac. Confirmed that cmake install doesn't do anything on the MSVC generator on Windows. I haven't checked to see what the Ninja generator does on windows, but this is probable fine.

**Pull request merged**